### PR TITLE
Added guide to install GHC and Cabal with a VS Code + HLS environment

### PR DESCRIPTION
**Pull request by thestaphmonster**

### CODE REVIEW

**Comment by tomjaguarpaw** [comment]
I tried this is Hyper V. Hyper V is causing me some problems, but I did get far enough to discover that the Windows binary distribution is an `.xz` file. My Windows installation (i.e. under Hyper V) didn't have any program that could understand that format. I notice that by contrast the cabal-install download is a `.zip` file. Does anyone have any ideas why the binary Windows GHC distribution wouldn't also be a `.zip` file for more easy compatibility with a default Windows install?

**Comment by MatthijsBlom** [comment]
> My Windows installation (i.e. under Hyper V) didn't have any program that could understand [the `.xz`] format.

I just used 7zip, twice.

**Comment by thestaphmonster** [comment]
@tomjaguarpaw .xz gives better compression. I did a test to see how much better:

image

Assuming that all the previous version of GHC are the same size as this one, given the 31 versions (>=7.6.2) currently hosted on haskell.org using this format, the amount of space saved is only 5GB. That's a very generous estimate because the size of GHC has increased dramatically since they started using this format.

Also, in theory, you should now be able to just use `tar` from the command line in Windows 10, but it seems to be [broken specifically for .xz](https://github.com/msys2/MSYS2-packages/issues/1548).

I could just add a dependency on 7zip to the guide in the meantime, but regardless, I do think the files should be .zip. I understand the premise that decreased disk space usage scales into the future, but the amount of space saved here's so small that it feels like the added friction for users isn't worth it, especially given the low (and continually decreasing) price of disk space.

File: Windows10.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+To add an item to your system path:
+* Type "environment" into your Windows search bar and select "Edit the system environment variables"
+* Click the "Environment Variables" button
+* Under "System variables", scroll down until you find "Path". Select it and click "Edit"
+* Click "New"
+* Enter the desired directory (e.g. `C:\Users\Will\Haskell\Cabal`) and hit OK
+* Enter the next desired directory (e.g `C:\Users\Will\Haskell\GHC\bin`) and hit OK
+* Hit "OK" on the "Environment Variables window"
+Note: If you don't hit "OK" on the last step it won't be saved, so our next test won't work
```
Comment (tomjaguarpaw) [inline Windows10.md:46]: After adding the "Cabal" entry, if you click "OK" it navigates you away from the dialog box. I think it's better to say "hit New" here.

The Windows UI for this is fiddly and *extremely* easy to get wrong in general.

File: Windows10.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+and
+
+```
+cabal-install version 3.2.0.0
+compiled using version 3.2.0.0 of the Cabal library
+```
+
+Run `cabal update`
```
Comment (tomjaguarpaw) [inline Windows10.md:68]: This is very easy to miss. We should try and find a way of making the actions that the user has to perform more prominent in general.

File: Windows10.md (hunk 1 above)
Comment (thestaphmonster) [reply Windows10.md:46]: You're right, that's a bug from when I was telling the user how to add a single item. I'll make the change.

**Comment by MatthijsBlom** [comment]
> [&hellip;] I did a test to see how much better:

I got the same results.

> Assuming that all the previous versions of GHC are the same size as this one [&hellip;]

GHC 9 is significantly smaller &ndash; or at least the `.xz` download is. I haven't yet tried zipping it.

> Also, in theory, you should now be able to just use `tar` from the command line in Windows 10, but it seems to be [broken specifically for .xz](https://github.com/msys2/MSYS2-packages/issues/1548).

For a few years or months now, `bsdtar` should be available by default on Windows. It is, however, unlikely to be found by those not in the know.

> I could just add a dependency on 7zip to the guide in the meantime, but regardless, I do think the files should be .zip. I understand the premise that decreased disk space usage scales into the future, but the amount of space saved here's so small that it feels like the added friction for users isn't worth it, especially given the low (and continually decreasing) price of disk space.

Agreed.

File: Windows10.md (hunk 2 above)
Comment (thestaphmonster) [reply Windows10.md:68]: Agreed, I'll give it its own header for now.

**Comment by tomjaguarpaw** [comment]
>.xz gives better compression.

Wow, that is really a lot better. Still, I think unavailability of `.zip` is big friction for Windows users.

**Comment by tomjaguarpaw** [comment]
> For a few years or months now, bsdtar should be available by default on Windows. It is, however, unlikely to be found by those not in the know.

Ah, this is very interesting. I'm not sure whether suggesting to use built-in bsdtar (which is awkward because it must be run from the command line) or suggesting to download 7-zip is the better course of action. Probably 7-zip is easier for people to use.

**Comment by tomjaguarpaw** [comment]
What role does MSYS2 play here? Does it provide the gcc toolchain that ghc uses to compile the code it generates?

**Comment by tomjaguarpaw** [comment]
Anyway, after making sure my Hyper-V had enough disk space, I can confirm these instructions work well under Hyper-V!

**Comment by thestaphmonster** [comment]
> What role does MSYS2 play here? Does it provide the gcc toolchain that ghc uses to compile the code it generates?

I actually have no concrete knowledge of this. From observation, it doesn't seem to be required for ghc, but for, I'm guessing, C libraries that certain Haskell packages depend upon.

So I think you can have a working ghc without MSYS2, but you'll run into issues with certain packages. Cairo is, apparently, a good example of one of those packages.

I don't like the fact that this appears necessary, though, and I find myself wondering what it'd take to make installing Haskell as easy as installing something like Rust. I don't know for certain, but a gut feeling tells me that maintaining an installer would be pretty awkward currently because of stuff like this. [This thread](https://gitlab.haskell.org/ghc/ghc/-/issues/18104) is a bit of evidence to confirm that suspicion.

> Anyway, after making sure my Hyper-V had enough disk space, I can confirm these instructions work well under Hyper-V!

That's great news, thanks for testing!

**Comment by tomjaguarpaw** [comment]
> I find myself wondering what it'd take to make installing Haskell as easy as installing something like Rust

*Most* of the instructions are essentially what ghcup does on Linux: download binary distributions of GHC and Cabal to a well-known location and update PATH as necessary.  The MSYS2 stuff makes it more complicated, though even on Linux the user has to ensure they have a C toolchain some other how.

This makes me think that "ghcup for Windows" would be possible and [@maerwald seems to confirm that suspicion on the thread](https://gitlab.haskell.org/ghc/ghc/-/issues/18104).

**Comment by MatthijsBlom** [comment]
> [&hellip;] I'm not sure whether suggesting to use built-in bsdtar (which is awkward because it must be run from the command line) or suggesting to download 7-zip is the better course of action. Probably 7-zip is easier for people to use.

Agreed on 7zip being probably easier.

For what it's worth:

-   `tar` is probably not quite universally present on Windows.
-   I tried `tar.exe -xf` on my machine; it hangs on `.xz` (huh?), but works fine with `.tar`. The WSL version works just fine, of course.

> What role does MSYS2 play here? Does it provide the gcc toolchain that ghc uses to compile the code it generates?

It provides `pkg-config`, at least. 

---

> [&hellip;] I find myself wondering what it'd take to make installing Haskell as easy as installing something like Rust. [&hellip;] [This thread](https://gitlab.haskell.org/ghc/ghc/-/issues/18104) [&hellip;]

I was hesitant before to mention it, on account of the (perceived) political situation, but having read the thread I feel a bit better, so here goes:

It looks to me like `stack` is quite viable as an installation option for beginners. It
-   just works, on all major platforms;
-   is widely used, so documentation, Q/A, and support are available;
-   provides such niceties as templates that can be made to fit beginners (#65).

Some downsides
-   slightly longer commands, or commands that do not correspond to those given in tutorials, i.e. `stack repl` vs `ghci`, `stack exec -- whatever`, &c;
-   Sometimes unfortunate compiler selection (as noted [by RAE](https://gitlab.haskell.org/ghc/ghc/-/issues/18104#note_280684)) (I wouldn't say this is an issue with stack, and moreover think it would be easily worked around _using_ stack.)
-   Bad interaction with system MSYS (as noted [by Ben Gamari](https://gitlab.haskell.org/ghc/ghc/-/issues/18104#note_291137)) (I never noticed this myself, and do not particularly expect beginners to either)
-   Lack of support for Backpack (as noted [by Tobias Haslop](https://gitlab.haskell.org/ghc/ghc/-/issues/18104#note_294628)) (I do not even know what Backpack is, and therefore do not expect beginners to need it.)
-   Risk of large disk space consumption.

**Comment by tomjaguarpaw** [comment]
Do you now how Stack interacts with HLS, particularly the default implementation of VSCode Haskell mode?

**Comment by MatthijsBlom** [comment]
No. I tried once, but over a year ago, and the situation was very different then. I made a separate issue to test this: #70.

Know, though, that WSL, by itself, provides a quite the hurdle as well.

File: Windows10.md (hunk 2 above)
Comment (MatthijsBlom) [reply Windows10.md:72]: Suggestion: replace the `cabal --version` test with the `cabal update`. Two birds with one stone.

Comment (tomjaguarpaw) [reply Windows10.md:72]: I think mentioning

```
ghc --version
cabal --version
cabal update
```

all at once might be the best combination. Then you can say you should see information like

* \<ghc version>
* \<cabal version>
* \<cabal update output>

**Comment by MatthijsBlom** [comment]
Whoops, I totally misread. Nothing to do with WSL, sorry.

Stack interacts with HLS just fine.

**Comment by thestaphmonster** [comment]
> I tried tar.exe -xf on my machine; it hangs on .xz (huh?), but works fine with .tar. The WSL version works just fine, of course.

This was my experience, too. Very odd.

> It looks to me like stack is quite viable as an installation option for beginners.

I recall trying to use Stack initially and I couldn't get it to work with everything I needed. I can't remember the details and it was quite a while ago, so it may not be the case anymore, but that was the reason I went with Cabal instead. I certainly don't have any objections to Stack over Cabal, this set of instructions was just the quickest route to a working installation that I could find. I'd be quite happy to switch to Stack if it made things simpler, though.

> This makes me think that "ghcup for Windows" would be possible and @maerwald seems to confirm that suspicion on the thread.

That's definitely along the lines of what I was thinking, minus the MSYS2 stuff, but I'd still be concerned with things like [this problem](https://github.com/haskell/haskell-language-server/issues/1372) that one of my initial testers ran into.

**Comment by thestaphmonster** [comment]
Just a note: I'm not deleting your pull request comments, they seem to be getting automatically deleted when I commit changes to the line that they're on.

**Comment by tomjaguarpaw** [comment]
Thanks for letting me know. GitHub seems to hide them from the diff when they get superseded.

**Comment by tomjaguarpaw** [comment]
@thestaphmonster It would be good to get this merged and then get some feedback from the wider public. Have a look at my latest PR into your branch. If and when you merge that we can have one more look at this PR and then merge it.

**Comment by thestaphmonster** [comment]
I merged your changes and added my name to the credits, ready to merge into master whenever you're happy.

**Comment by tomjaguarpaw** [comment]
Fantastic. I merged it. Here it is: https://github.com/tomjaguarpaw/tilapia/blob/master/Windows10.md

It would be good to get feedback from some Windows users. Perhaps we should make a post on Reddit explaining that you have written this guide and asking for feedback about how easy it is to follow and whether anything else needs addressing. Would you like to make such a post?

**Comment by thestaphmonster** [comment]
Excellent! And yes that's a good idea, I should get a chance to make a post tomorrow.

**Comment by thestaphmonster** [comment]
I've made a thread: https://www.reddit.com/r/haskell/comments/lur2hn/a_simple_guide_setup_a_haskell_development/

Hopefully it'll go smoothly for most users!

**Comment by tomjaguarpaw** [comment]
We should think about where to take this next. I would suggest we tweet about it and publish it on the Haskell Discourse. After that I'll try to raise awareness of it amongst people who are working on similar things, such as a potential Windows installer.

**Comment by tomjaguarpaw** [comment]
See [some discussion about a possible Windows installer](https://haskell-foundation.slack.com/archives/C01PKJ9QKJ8/p1615049973008100) on the Haskell Foundation Slack.

**Comment by thestaphmonster** [comment]
> I would suggest we tweet about it and publish it on the Haskell Discourse.

I'm not a Twitter user I'm afraid! Would that be something you'd be happy to do?

Also, a message I received on Reddit has highlighted a problem with the Chocolatey installation route. 

They were getting the error message `haskell-language-server 1.0.0 for GHC 9.0.1 is not available on Windows_NT` when they got to the VS Code section. The problem was rectified by following the manual installation instructions and using GHC 8.10.3. 

It seems to me that our options to prevent that from happening in the future are:
- Remove the Chocolatey section
- Communicate with whoever's responsible for the Chocolatey package and the HLS team to see what can be done; whether or not they can make sure that they stay in sync so that the default version on Chocolatey is always HLS compatible. Something like that
- Specify a known-working version for Chocolatey to download. I don't think this one is actually possible because the haskell-dev package seems to just be a single version, with the script it runs presumably pulling the latest compatible version of GHC

Additionally, we had [a new post in the Reddit thread](https://www.reddit.com/r/haskell/comments/lur2hn/a_simple_guide_setup_a_haskell_development/gq3lmdr/?context=3) which contains some good suggestions. The problem is, the recommended GHC 8.10.4 didn't work for the first of my test subjects; if I recall correctly, HLS didn't support it at the time. Maybe that's changed now. With regards to the path length stuff, I'm a bit on the fence as to whether that should be addressed by this guide given its already excessive length.

I was having thoughts as to how we could potentially have a web of information surrounding new user onboarding. Perhaps the path length stuff recommended here could be in another document that's mentioned as a kind of "hey, you might have the problems x, y and z, addressed here, here and here" type thing.

> See some discussion about a possible Windows installer on the Haskell Foundation Slack.

I've not got an account on there and I seem to be unable to create one without speaking to a workspace administrator. Do you know who I could contact to get one?

**Comment by tomjaguarpaw** [comment]
> > Twitter
>
>  Would that be something you'd be happy to do?

Yes indeed


> > HF Slack
>
> I've not got an account on there and I seem to be unable to create one without speaking to a workspace administrator. Do you know who I could contact to get one?

Anyone is welcome to create an account through the [public invite](https://join.slack.com/t/haskell-foundation/shared_invite/zt-mjh76fw0-CEjg2NbyVE8rVQDvR~0F4A). For technical reasons I don't think it's possible to make the main sign up page accessible to everyone.

**Comment by tomjaguarpaw** [comment]
I have tweeted: https://twitter.com/tomjaguarpaw/status/1371155128812773381

**Comment by thestaphmonster** [comment]
The chocolatey section is going to keep tripping people up given the latency between GHC updates and HLS support.

Any objection to just removing that section altogether, for the time being?

**Comment by tomjaguarpaw** [comment]
If you think that's best then go for it! This could even be a "manual install guide" i.e. deliberately have nothing to say about chocolatey.

**Comment by thestaphmonster** [comment]
Great, I've removed it in https://github.com/tomjaguarpaw/tilapia/pull/84

**Comment by thestaphmonster** [comment]
I made a discourse thread: https://discourse.haskell.org/t/a-simple-guide-set-up-a-haskell-development-environment-in-windows-10/2194

**Comment by thestaphmonster** [comment]
@tomjaguarpaw for a small change like your suggestion in that discourse thread, would you prefer a change to just be pushed straight to master or would you prefer a new branch with a pull request and an approval for all changes?

Only asking to avoid annoying you with pull requests for tiny changes if that's not what you want!

**Comment by tomjaguarpaw** [comment]
I'm happy to receive small pull requests. In fact the smaller the better! I generally find it very quick and easy to review small PRs.

**Pull request merged**